logsmanagement/db_api.c

```diff
@@ -0,0 +1,1096 @@
+		
+	    freez(temp_msg.text_compressed);
+        
+        if(p_query_params->results_buff->len >= p_query_params->quota){
+        	p_query_params->end_timestamp = temp_msg.timestamp;
+        	break;
+        }
+
+        rc = sqlite3_step(stmt_retrieve_log_msg_metadata);
+        // debug(D_LOGS_MANAG, "Query: %s\n rc:%d\n", sqlite3_expanded_sql(stmt_retrieve_log_msg_metadata), rc);
+        if (unlikely(rc != SQLITE_ROW && rc != SQLITE_DONE)) fatal_sqlite3_err(rc, __LINE__);
+    }
+
+    rc = sqlite3_exec(p_file_info->db, "END TRANSACTION;", NULL, NULL, NULL);
+    if (unlikely(rc != SQLITE_OK)) fatal_sqlite3_err(rc, __LINE__);
+    rc = sqlite3_finalize(stmt_retrieve_log_msg_metadata);
+    if (unlikely(rc != SQLITE_OK)) fatal_sqlite3_err(rc, __LINE__);
+}
+
+#define TMP_VIEW_TABLE		   "compound_view"
+#define TMP_VIEW_QUERY_PREFIX  "CREATE TEMP VIEW " TMP_VIEW_TABLE " AS SELECT * FROM (SELECT * FROM '0'."\
+							    LOGS_TABLE " INNER JOIN (VALUES(0)) ORDER BY Timestamp) "
+#define TMP_VIEW_QUERY_BODY_1  "UNION ALL SELECT * FROM (SELECT * FROM '"
+#define TMP_VIEW_QUERY_BODY_2  "'." LOGS_TABLE " INNER JOIN (VALUES("
+#define TMP_VIEW_QUERY_BODY_3  ")) ORDER BY Timestamp) "
+#define TMP_VIEW_QUERY_POSTFIX "ORDER BY Timestamp;" 
+void db_search_compound(logs_query_params_t *const p_query_params, struct File_info *const p_file_infos[]) {
+	int rc = 0;
+	int pfi_off = 0;
+
+	sqlite3 *dbt;
+	rc = sqlite3_open_v2(p_file_infos[0]->db_metadata, &dbt, SQLITE_OPEN_READONLY, NULL);
+	if (unlikely(rc != SQLITE_OK)) fatal_sqlite3_err(rc, __LINE__);
+
+	rc = sqlite3_exec(dbt, "BEGIN TRANSACTION;", NULL, NULL, NULL);
+    if (unlikely(rc != SQLITE_OK)) fatal_sqlite3_err(rc, __LINE__);	
+
+	/* Attach DBs */
+	sqlite3_stmt *stmt_attach_db;
+	rc = sqlite3_prepare_v2(dbt,"ATTACH DATABASE ?  AS ? ;", -1, &stmt_attach_db, NULL);
+	if (unlikely(rc != SQLITE_OK)) fatal_sqlite3_err(rc, __LINE__);
+	for(pfi_off = 0; p_file_infos[pfi_off]; pfi_off++){
+		rc = sqlite3_bind_text(stmt_attach_db, 1, p_file_infos[pfi_off]->db_metadata, -1, NULL);
+		// rc = sqlite3_bind_text(stmt_attach_db, 2, METADATA_DB_FILENAME, -1, NULL);
+		if (unlikely(rc != SQLITE_OK)) fatal_sqlite3_err(rc, __LINE__);
+		rc = sqlite3_bind_int(stmt_attach_db, 2, pfi_off);
+		if (unlikely(rc != SQLITE_OK)) fatal_sqlite3_err(rc, __LINE__);
+		// debug(D_LOGS_MANAG, "Attach db query[%d]:%s", pfi_off, sqlite3_expanded_sql(stmt_attach_db));
+		rc = sqlite3_step(stmt_attach_db);
+		if (unlikely(rc != SQLITE_DONE)) fatal_sqlite3_err(rc, __LINE__);
+		rc = sqlite3_reset(stmt_attach_db);
+		if (unlikely(rc != SQLITE_OK)) fatal_sqlite3_err(rc, __LINE__);
+	}
+
+	/* Create temporary view */
+	char tmp_view_query[sizeof(TMP_VIEW_QUERY_PREFIX) + (
+							sizeof(TMP_VIEW_QUERY_BODY_1) + 
+						 	sizeof(TMP_VIEW_QUERY_BODY_2) + 
+						 	sizeof(TMP_VIEW_QUERY_BODY_3) + 4 ) * (MAX_COMPOUND_QUERY_SOURCES - 1) + 
+						sizeof(TMP_VIEW_QUERY_POSTFIX) + 50] = TMP_VIEW_QUERY_PREFIX; // +50 bytes - play it safe
```

Comment:
Why is `50` the safest and not `49` or `51`? :)
This is another `magic number` that needs a `define`.

Reply:
I don't think this falls under the same case as with the others. In this case, 50 is really a magic number, that's why the comment :). Could be 49, could be 51, just some extra bytes to play it safe :).

Reply:
it is really strange this number :thinking:. I will think more about this.